### PR TITLE
Backport authorization procs to 1.x branch

### PR DESCRIPTION
**Pull request by jarl-dk**
## Description

This is a backport  of authorization using `Proc`s. It is a simple cherry-pick from `main` branch.

(cherry picked from commit fe600c72adc79180ad7e545191edd5dd2fe61990)

## Todos
List any remaining work that needs to be done, i.e:
- [x] Tests
- [x] Rubocop
- [x] Documentation

## Additional Notes
This is a backport  of authorization using `Proc`s. It is a simple cherry-pick from `main` branch.
(cherry picked from commit fe600c72adc79180ad7e545191edd5dd2fe61990)

### CODE REVIEW

**Comment by olleolleolle** [comment]
@jarl-dk Thanks for taking a look at this. 

A blocing issue with this cherry-pick is that it takes the step of removing the feature from the 1.x release series, which we only want to fully remove in the 2.x, which is brewing.

**Comment by jarl-dk** [comment]
@olleolleolle : Yes, I just realised this myself. Thanks. I'll see if I can make a backward compatible fixup (selective cherry-pick).

**Comment by jarl-dk** [comment]
The PR does contains some bloat code (if just reaching for `Proc` support). But my aim has been minimal code diff to `v1.7.1` and as close as possible to `fe600c72adc79180ad7e545191edd5dd2fe61990`.

**Comment by iMacTia** [comment]
Thank you @jarl-dk, indeed I can see the extra code for basic auth support in the generic `Authorization` middleware.
I'm happy to clean up myself after this PR if you don't have the time, but I'd really like to see the new tests ported as well and the existing ones still passing before merging this.
There also seems to be a small issue with linting that will need to be fixed before merging.

**Comment by jarl-dk** [comment]
I'll give it a shot more...

File: lib/faraday/request/authorization.rb [hunk 1]
```diff
@@ -39,16 +41,53 @@ def self.build_hash(type, hash)
...
 
       # @param env [Faraday::Env]
-      def call(env)
-        env.request_headers[KEY] = @header_value unless env.request_headers[KEY]
-        @app.call(env)
+      # def call(env)
+      #   env.request_headers[KEY] = @header_value unless env.request_headers[KEY]
+      #   @app.call(env)
+      # end
+      #
```
Comment (partydrone) [inline lib/faraday/request/authorization.rb:60]: Perhaps this will fix the Yard junk error?

```suggestion

```

**Review by partydrone — CHANGES_REQUESTED** [review]

File: lib/faraday/request/authorization.rb (hunk 1 above)
Comment (iMacTia) [reply lib/faraday/request/authorization.rb:60]: Ah yes, YARD thinks this is a single comment for the `on_request` method.
@jarl-dk It doesn't make sense to leave the old `call` method commented, we can simply get rid of it.

Comment (jarl-dk) [reply lib/faraday/request/authorization.rb:60]: I agree that was a mistake, I'll remove it.

**Comment by jarl-dk** [comment]
I have updated the PR to not remove any tests and just add the test regarding Proc

**Review by iMacTia — APPROVED** [review]
This is good enough, thanks @jarl-dk, I'll take over from here for the cleanup 👍

**Pull request merged**